src/coreclr/gc/gc.cpp

```diff
@@ -25285,49 +25389,306 @@ int gc_heap::calculate_new_heap_count ()
         {
-            dprintf (6666, ("[CHP1] stcp %.3f > 5, %d + %d = %d", smoothed_median_throughput_cost_percent, n_heaps, step_up, (n_heaps + step_up)));
+            float step_up_percent = log_with_base((smoothed_median_throughput_cost_percent - target_tcp + log_base), log_base);
+            float step_up_float = (float)(step_up_percent / 100.0 * actual_n_max_heaps);
+            int step_up_int = (int)step_up_float;
+
+            dprintf (6666, ("[CHP0] inc %d(%.3f), last inc %d, %Id GCs elapsed, last stcp %.3f",
+                step_up_int, step_up_float, (int)dynamic_heap_count_data.last_changed_count,
+                (current_gc_index - dynamic_heap_count_data.last_changed_gc_index), dynamic_heap_count_data.last_changed_stcp));
+
+            // Don't adjust if we just adjusted last time we checked, unless we are in an extreme situation.
+            if ((smoothed_median_throughput_cost_percent < 20.0f) &&
+                (avg_throughput_cost_percent < 20.0f) &&
+                ((current_gc_index - dynamic_heap_count_data.last_changed_gc_index) < (2 * dynamic_heap_count_data_t::sample_size)))
+            {
+                dprintf (6666, ("[CHP0] we just adjusted %Id GCs ago, skipping", (current_gc_index - dynamic_heap_count_data.last_changed_gc_index)));
+            }
+            else
+            {
+                if (step_up_int)
+                {
+                    if (dynamic_heap_count_data.dec_failure_count)
+                    {
+                        dprintf (6666, ("[CHP0] intending to grow, reset dec failure count (was %d)", dynamic_heap_count_data.dec_failure_count));
+                        dynamic_heap_count_data.dec_failure_count = 0;
+                    }
+
+                    if (((int)dynamic_heap_count_data.last_changed_count > 0) && (dynamic_heap_count_data.last_changed_gc_index > 0.0) &&
+                        ((current_gc_index - dynamic_heap_count_data.last_changed_gc_index) <= (3 * dynamic_heap_count_data_t::sample_size)))
+                    {
+                        dprintf (6666, ("[CHP0-0] just grew %d GCs ago, no change", (current_gc_index - dynamic_heap_count_data.last_changed_gc_index)));
+                        step_up_int = 0;
+                    }
+                    else
+                    {
+                        // If the calculation tells us to grow, we should check to see if the slope has been coming down rapidly, if so there's no reason to grow.
+                        int above_target_tcp_count = dynamic_heap_count_data.rearrange_recorded_tcp ();
+                        float above_target_tcp_slope = slope (dynamic_heap_count_data.recorded_tcp_rearranged, above_target_tcp_count, &avg_recorded_tcp);
+                        float diff_pct = (target_tcp - avg_recorded_tcp) / target_tcp;
+                        float adjusted_target_tcp = target_tcp * 1.2f;
```

Comment:
not sure I follow why its 1.2x here?

Reply:
+/- 20% is treated as in range of target. I've moved this to the header.